WorkflowConcurrency/Tests/WorkerTests.swift

```diff
@@ -0,0 +1,204 @@
+        )
+
+        let expectation = XCTestExpectation()
+        let disposable = host.rendering.signal.observeValues { rendering in
+            expectation.fulfill()
+        }
+
+        XCTAssertEqual(0, host.rendering.value)
+
+        wait(for: [expectation], timeout: 5.0)
+        XCTAssertEqual(1, host.rendering.value)
+
+        disposable?.dispose()
+    }
+
+    func testWorkerOutput() {
+        let host = WorkflowHost(
+            workflow: TaskTestWorkerWorkflow(key: "")
+        )
+
+        let expectation = XCTestExpectation()
+        let disposable = host.rendering.signal.observeValues { rendering in
+            expectation.fulfill()
+        }
+
+        XCTAssertEqual(0, host.rendering.value)
+
+        wait(for: [expectation], timeout: 5.0)
+        XCTAssertEqual(1, host.rendering.value)
+
+        disposable?.dispose()
+    }
+
+    func testExpectedWorker() {
+        TaskTestWorkerWorkflow(key: "123")
+            .renderTester()
+            .expectWorkflow(
+                type: WorkerWorkflow<TaskTestWorker>.self,
+                key: "123",
+                producingRendering: (),
+                producingOutput: 1,
+                assertions: { _ in }
+            )
+            .render { _ in }
+            .verifyState { state in
+                XCTAssertEqual(state, 1)
+            }
+    }
+
+    // A worker declared on a first `render` pass that is not on a subsequent should have the work cancelled.
+    func test_cancelsWorkers() {
+        struct WorkerWorkflow: Workflow {
+            typealias State = Void
+
+            enum Mode {
+                case notWorking
+                case working(start: XCTestExpectation, end: XCTestExpectation)
+            }
+
+            let mode: Mode
+
+            func render(state: State, context: RenderContext<WorkerWorkflow>) -> Bool {
+                switch mode {
+                case .notWorking:
+                    return false
+                case .working(start: let startExpectation, end: let endExpectation):
+                    ExpectingWorker(
+                        startExpectation: startExpectation,
+                        endExpectation: endExpectation
+                    )
+                    .mapOutput { _ in AnyWorkflowAction.noAction }
+                    .running(in: context)
+                    return true
+                }
+            }
+
+            struct ExpectingWorker: Worker {
+                typealias Output = Void
+
+                let startExpectation: XCTestExpectation
+                let endExpectation: XCTestExpectation
+
+                func run() async -> Void {
+                    startExpectation.fulfill()
+                    for _ in 1 ... 4 {
+                        if Task.isCancelled {
+                            endExpectation.fulfill()
+                            return
+                        }
+                        try? await Task.sleep(nanoseconds: 500000000)
```

Comment:
Can we figure out a way to test this without an explicit sleep?

Reply:
We don't control what thread/scheduler the async Task runs on so I don't know of a way of moving the scheduler forward in time for this.  The test needs the async function to be busy longer than the timeout so that we can verify cancellation actually works.  We have to cancel out of work that would normally go longer than the timeout.
The way the test is currently coded it should only wait for the first sleep so I can lower that to something very quick and try that so that the test runs faster.